File: internal/service/logic_engine_parity_test.go
package service

import (
	"encoding/json"
	"os"
	"testing"

	"github.com/ahmadasror/txsurvey/internal/model"
)

// TestLogicEngineParity is the Go half of the DUAL-ENGINE PARITY GATE. It runs the
// authoritative Go conditionMatches over the SAME fixture the TS engine test reads
// (frontend/src/lib/logicEngine.parity.test.ts). If the two engines drift, one of
// the two sides goes red. See docs/architecture/adr/001-dual-logic-engine.md.
func TestLogicEngineParity(t *testing.T) {
	raw, err := os.ReadFile("testdata/logic_parity_cases.json")
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var fixture struct {
		Cases []struct {
			Name         string              `json:"name"`
			Answer       json.RawMessage     `json:"answer"`
			Operator     model.LogicOperator `json:"operator"`
			CompareValue json.RawMessage     `json:"compareValue"`
			Expected     bool                `json:"expected"`
		} `json:"cases"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatalf("parse fixture: %v", err)
	}
	if len(fixture.Cases) == 0 {
		t.Fatal("fixture has no cases")
	}
	for _, c := range fixture.Cases {
		t.Run(c.Name, func(t *testing.T) {
			if !model.ValidLogicOperator(c.Operator) {
				t.Fatalf("fixture uses unknown operator %q", c.Operator)
			}
			if got := conditionMatches(c.Answer, c.Operator, c.CompareValue); got != c.Expected {
				t.Fatalf("conditionMatches(%s, %s, %s) = %v, want %v",
					c.Answer, c.Operator, c.CompareValue, got, c.Expected)
			}
		})
	}
}

File: frontend/src/lib/logicEngine.parity.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { conditionMatches } from "./logicEngine";
import type { AnswerValue, LogicOperator } from "@/types/forms";

// TS half of the DUAL-ENGINE PARITY GATE. Reads the SAME fixture the Go test reads
// (internal/service/testdata/logic_parity_cases.json) and asserts the TS engine
// agrees with `expected` on every row. If logic_engine.go and logicEngine.ts drift
// apart, one of the two suites goes red. See docs/architecture/adr/001-dual-logic-engine.md.
interface ParityCase {
  name: string;
  answer: AnswerValue | null;
  operator: LogicOperator;
  compareValue: unknown;
  expected: boolean;
}

// frontend/src/lib -> repo root -> internal/service/testdata/...
const fixturePath = fileURLToPath(
  new URL("../../../internal/service/testdata/logic_parity_cases.json", import.meta.url),
);
const fixture = JSON.parse(readFileSync(fixturePath, "utf8")) as { cases: ParityCase[] };

describe("dual-engine parity (TS side of the shared fixture)", () => {
  it("has a non-empty fixture", () => {
    expect(fixture.cases.length).toBeGreaterThan(0);
  });

  for (const c of fixture.cases) {
    it(c.name, () => {
      const answer = (c.answer ?? undefined) as AnswerValue | undefined;
      expect(conditionMatches(answer, c.operator, c.compareValue)).toBe(c.expected);
    });
  }
});
